### ISSUE TITLE
Dynamic mutator and accessor support

### ISSUE THREAD

**Issue by brendt** [laravel/ideas#1902]
Right now, mutators and accessors only take methods with a specific naming convention into account, eg:

```php
// \Illuminate\Database\Eloquent\Concerns\HasAttributes::mutateAttribute

protected function mutateAttribute($key, $value)
{
    return $this->{'get'.Str::studly($key).'Attribute'}($value);
}
```

My suggestion is to keep this default behaviour, but add another way of registering accessors and mutators on top of it. This would allow packages to add richer model behaviour. Here's an example.

I maintain a package which adds state behaviour to models. A user can register fields to be state fields, and the package will convert their stored string value to state objects:

```php
/**
 * @property \App\States\PaymentState state
 */
class Payment extends Model
{
    use HasStates;

    protected function registerStates(): void
    {
        $this
            ->addState('state', PaymentState::class)
            ->allowTransition(Pending::class, Paid::class)
            ->allowTransition(Pending::class, Failed::class, PendingToFailed::class);
    }
}
```

```php
$payment->state = new PendingState();

// The package will take care of serialising and deserialising the state from and to the database
```

One feature that's still missing, is that the state object doesn't know what field it belongs to, hence making some methods more complex and more difficult to use. I've got a working solution though, by overriding `__set` in my `HasStates` trait, you can take a look at it here: https://github.com/spatie/laravel-model-states/pull/31

The issue though, is that if this model uses another trait which overrides `__set`, we're in trouble. The same goes for `setAttribute()`.

My suggestion is to allow several mutators and accessors per field, like so:

```php
public function hasGetMutator($key)
{
    return $this->hasGetMutatorMethod($key) || isset($this->mutators[$key]);
}

public function hasGetMutatorMethod($key)
{
    return method_exists($this, 'get'.Str::studly($key).'Attribute');
}

protected function mutateAttribute($key, $value)
{
    if ($this->hasGetMutatorMethod($key)) {
        return $this->{'get'.Str::studly($key).'Attribute'}($value);
    }

    foreach ($this->mutators[$key] ?? [] as $mutator) {
        $value = $mutator($value);
    }

    return $value;
}
```

Obviously we could discuss naming and all the details, this is a simple quick example to show what this idea is about.

In summary: this feature would allow packages to add richer model functionality, without having to worry about their traits colliding with other packages.

**Comment by brendt** [comment]
Can I get some input on this? @driesvints @taylorotwell ? Depending on whether you'd consider adding this feature in the framework, I need to rework a package before tagging a major release.

**Comment by driesvints** [comment]
How do you define the sequence? In your example, obviously the mutator on the model itself takes precedence and is the only one executed but the foreach loops just loops over different mutators. What will happen when mutators from different packages collide?

At the moment it would just seem to me that it'd be best to always be explicit and let the end user define the sequence for mutating in the model mutator? (everything above also applies to accessors)

**Comment by brendt** [comment]
How I envisioned it was that traits could add dynamic mutators in their unique boot method:

```php
public function bootMyTrait() {
    self::registerMutator('fieldName', fn () => …);
}
```

I don't think i'd happen very often that several traits would act on the same field, though if that happens they would simply be executed in the order of the traits.

**Comment by driesvints** [comment]
@brendt Taylor let me know you're free to send in a PR so he can have a look at how you'd tackle it :)

Make sure to give a thorough explanation on the pr. Thanks!

**Comment by brendt** [comment]
Since Taylor is thinking about adding custom casts, I guess this issue isn't relevant anymore (https://twitter.com/taylorotwell/status/1213652671611232257)
